File: moqt/internal/message/goaway.go
package message

import (
	"io"
)

/*
 *	GOAWAY Message {
 *	  Message Length (i)
 *	  New Session URI (s)
 *	}
 */
type GoawayMessage struct {
	NewSessionURI string
}

func (gm GoawayMessage) Len() int {
	return StringLen(gm.NewSessionURI)
}

func (gm GoawayMessage) Encode(w io.Writer) error {
	var err error
	msgLen := gm.Len()
	b := make([]byte, 0, msgLen+VarintLen(uint64(msgLen)))

	b, _, err = WriteMessageLength(b, uint64(msgLen))
	if err != nil {
		return err
	}
	b, _, err = WriteString(b, gm.NewSessionURI)
	if err != nil {
		return err
	}

	_, err = w.Write(b)

	return err
}

func (gm *GoawayMessage) Decode(src io.Reader) error {
	num, err := ReadMessageLength(src)
	if err != nil {
		return err
	}

	if num > MaxMessageSize {
		return ErrMessageTooLarge
	}

	b := make([]byte, num)

	_, err = io.ReadFull(src, b)
	if err != nil {
		return err
	}

	str, n, err := ReadString(b)
	if err != nil {
		return err
	}
	gm.NewSessionURI = str
	b = b[n:]

	if len(b) != 0 {
		return ErrMessageTooShort
	}

	return nil
}

File: moq-web/src/internal/message/goaway.ts
import type { Reader, Writer } from "@okdaichi/golikejs/io";
import { MessageDecoder, MessageEncoder, readFull, readVarint } from "./message.ts";

export interface GoawayMessageInit {
	newSessionURI?: string;
}

export class GoawayMessage {
	newSessionURI: string;

	constructor(init: GoawayMessageInit = {}) {
		this.newSessionURI = init.newSessionURI ?? "";
	}

	async encode(w: Writer): Promise<Error | undefined> {
		const e = new MessageEncoder();
		e.string(this.newSessionURI);
		const [, err] = await w.write(e.frame());
		return err;
	}

	async decode(r: Reader): Promise<Error | undefined> {
		const [msgLen, , err1] = await readVarint(r);
		if (err1) return err1;

		const buf = new Uint8Array(msgLen);
		const [, err2] = await readFull(r, buf);
		if (err2) return err2;

		const d = new MessageDecoder(buf);
		this.newSessionURI = d.string();

		return undefined;
	}
}
